File: Python/0128-h-longest-consecutive-sequence.py
"""
Primitive idea
Sorting
Time: O(nlogn)
Space: O(n)
"""
class Solution_1:
    def longestConsecutive(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        nums = sorted(nums)
        count = 1
        maxCount = 1
        for i in range(1, len(nums)):
            if nums[i] == nums[i-1] + 1:
                count += 1
            else:
                count = 1

            maxCount = max(count, maxCount)

        return maxCount


"""
Set
Put nums into set
For each element of nums, find its lower and upper bound
Time: O(n)
Space: O(n)
"""
class Solution_2:
    def longestConsecutive(self, nums: List[int]) -> int:
        if not nums:
            return 0

        numSet = set(nums)
        for item in nums:
            numSet.add(item)

        res = 0
        for item in nums:
            if item in numSet:
                numSet.remove(item)

                l = item - 1
                r = item + 1
                while l in numSet:
                    numSet.remove(l)
                    l -= 1
                while r in numSet:
                    numSet.remove(r)
                    r += 1
                res = max(res, r - l - 1)
        return res

File: leetcode/C++/0128-h-longest-consecutive-sequence.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        if (!nums.size())
            return 0;
        
        unordered_map<int, int> map;
        int l = 1;
        for (auto i : nums)
        {
            if (map.find(i) != map.end()) continue;
            map[i] = 1;
            if (map.find(i - 1) != map.end())
                l = max(l, mergeCluster(map, i - 1, i));
            if (map.find(i + 1) != map.end())
                l = max(l, mergeCluster(map, i, i + 1));
        }
        return l;
    }

private:
    int mergeCluster(unordered_map<int, int> &map, int left, int right)
    {
        int upper = right + map[right] - 1;
        int lower = left - map[left] + 1;
        int length = upper - lower + 1;
        map[upper] = length;
        map[lower] = length;
        return length;
    }
};